xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,373 @@
+	// RPC, and thus should use the most updated form of the current balancer.
+	gsb.cc.UpdateState(state)
+}
+
+// swap swaps out the current lb with the pending LB and updates the ClientConn.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) swap() {
+	gsb.cc.UpdateState(gsb.pendingState)
+	currBalToClose := gsb.balancerCurrent
+	gsb.balancerCurrent = gsb.balancerPending
+	gsb.balancerPending = nil
+	for sc, bal := range gsb.scToSubBalancer {
+		if bal == currBalToClose {
+			gsb.cc.RemoveSubConn(sc)
+			delete(gsb.scToSubBalancer, sc)
+		}
+	}
+	go func() {
+		gsb.currentMu.Lock()
+		defer gsb.currentMu.Unlock()
+		currBalToClose.Close()
+	}()
+}
+
+// Helper function that checks if the balancer passed in is current or pending.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) balancerCurrentOrPending(bal balancer.Balancer) bool {
+	return bal == gsb.balancerCurrent || bal == gsb.balancerPending
+}
+
+func (gsb *gracefulSwitchBalancer) newSubConn(bal balancer.Balancer, addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return nil, fmt.Errorf("%T at address %p that called NewSubConn is deleted", bal, bal)
+	}
+	sc, err := gsb.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		return nil, err
+	}
+	gsb.scToSubBalancer[sc] = bal
+	return sc, nil
+}
+
+func (gsb *gracefulSwitchBalancer) resolveNow(bal balancer.Balancer, opts resolver.ResolveNowOptions) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+
+	// If the resolver specifies an update with a whole new type of policy,
+	// there is no reason to forward a ResolveNow call from the balancer being
+	// switched from, as the most recent update from the Resolver does not
+	// concern the balancer being switched from.
+	if bal == gsb.balancerCurrent && gsb.balancerPending != nil {
+		return
+	}
+	gsb.cc.ResolveNow(opts)
+}
+
+func (gsb *gracefulSwitchBalancer) removeSubConn(bal balancer.Balancer, sc balancer.SubConn) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+	delete(gsb.scToSubBalancer, sc)
+	gsb.cc.RemoveSubConn(sc)
+}
+
+func (gsb *gracefulSwitchBalancer) updateAddresses(bal balancer.Balancer, sc balancer.SubConn, addrs []resolver.Address) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+	gsb.cc.UpdateAddresses(sc, addrs)
+}
+
+// SwitchTo gracefully switches to the new balancer. This function must be
+// called synchronously alongside the rest of the balancer.Balancer methods this
+// Graceful Switch Balancer implements.
+func (gsb *gracefulSwitchBalancer) SwitchTo(builder balancer.Builder) error {
+	gsb.mu.Lock()
+	if gsb.closed {
+		gsb.mu.Unlock()
+		// logger?
+		return errBalancerClosed
+	}
+	ccw := &clientConnWrapper{
+		gsb: gsb,
+	}
+	newBalancer := builder.Build(ccw, gsb.bOpts)
```

Comment:
Hmm, what if `Build()` sends a picker update? It will block on `gsb.mu`.

And even if it doesn't block, the update will be dropped, because `gsb.balancerPending` doesn't know this balancer (fields not updated yet).

Reply:
I read every single Build() method in the codebase, and couldn't find an inline Picker update. Yeah, if it did have an inline Picker update, it would be wrong because a. it would block, and b. wouldn't be current or pending, and wouldn't even be processed. Also, this is how c does it (builds the balancer than writes to current/pending, which is used in UpdateState() to determine if should process the update), thus they don't have an inline Picker update from build, correct? Also, if I'm right about no inline callback, it doesn't seem like something we will support in the future. Also, the point where this utility struct won't have children that are the child nodes in the full lb tree, and thus will always require extra configuration from somewhere before calling back into UpdateState()?

Reply:
https://en.wikipedia.org/wiki/Proof_by_example

We most definitely should consider that `Build` could call `UpdateState`.  There's nothing in the API documentation that forbids it, and it would work if the child policy was the direct LB policy of the channel.  I'm not sure about C (is this an API constraint, an "oversight", or a "known limitation"), but I do know that their API is private for now, so they only need to support their own LB policies.

For us, could we do something like set a field in `clientConnWrapper` to indicate the policy is currently being constructed?  (This suggestion is definitely following along the path of "adding stuff to make a corner case work" when it may be possible to "find a better representation" and avoid the need for this special case entirely.)

Reply:
I think I came up with a correct solution. Let me know if it's incorrect/bloated (too much code/logic and if there's a simpler solution.